File: 6. Graph/12_Kahn_Algo.cpp
#include <bits/stdc++.h>
using namespace std;

bool iscyclic(int v, vector<int> adj[])
{
    vector<int> indeg(v, 0);
    for (int i = 0; i < v; i++)
    {
        for (int j : adj[i])
            indeg[j]++;
    }

    queue<int> q;
    int ans = 0;
    unordered_set<int> vis;

    for (int i = 0; i < v; i++)
    {
        if (indeg[i] == 0)
        {
            q.push(i);
            ans += 1;
        }
    }

    while (!q.empty())
    {
        int currvertex = q.front();
        q.pop();
        if (vis.find(currvertex) != vis.end())
            continue;

        vis.insert(currvertex);
        for (int neighbour : adj[currvertex])
        {
            indeg[neighbour] -= 1;
            if (indeg[neighbour] == 0)
            {
                q.push(neighbour);
                ans += 1;
            }
        }
    }
    if (ans == v)
        return false;
    return true;
}

File: DSA/6. Graph/12_Kahn_Algo.java
/*
    Problem Link: https://www.geeksforgeeks.org/detect-cycle-in-a-directed-graph-using-bfs/
    
    Kahn's Algorithm - Detect Cycle in Directed Graph using BFS
*/

import java.util.*;

class Kahn_Algo {
    public static boolean iscyclic(int v, List<Integer>[] adj) {
        int[] indeg = new int[v];
        for (int i = 0; i < v; i++) {
            for (int j : adj[i]) {
                indeg[j]++;
            }
        }

        Queue<Integer> q = new LinkedList<>();
        int ans = 0;
        Set<Integer> vis = new HashSet<>();

        for (int i = 0; i < v; i++) {
            if (indeg[i] == 0) {
                q.offer(i);
                ans += 1;
            }
        }

        while (!q.isEmpty()) {
            int currvertex = q.poll();
            if (vis.contains(currvertex)) {
                continue;
            }

            vis.add(currvertex);
            for (int neighbour : adj[currvertex]) {
                indeg[neighbour] -= 1;
                if (indeg[neighbour] == 0) {
                    q.offer(neighbour);
                    ans += 1;
                }
            }
        }
        if (ans == v) {
            return false;
        }
        return true;
    }
}
